HeterogeneousCore/AlpakaTest/test/BuildFile.xml

```diff
@@ -1,4 +1,4 @@
 <test
   name="testHeterogeneousCoreAlpakaTestWriteRead"
-  command="cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py; cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py"
+  command="mkdir -p testHeterogeneousCoreAlpakaTestWriteRead; cd testHeterogeneousCoreAlpakaTestWriteRead; cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py; cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py"
```

Comment:
Should this line use `&&` insread of `;` in order to proceed only if the preceding command succeeded? @smuzaffar

Reply:
@makortel, yes please use `&&` instead of `;` . Note that you need to use `&amp;&amp;` so that it can be properly parsed.

Reply:
@fwyzard , I do not think xml parser is going to parse it. you need to use something like
```
<test
   name="NAME"
   command="cmd arg1
   arg2
   arg3"
/>
```

Reply:
The current syntax does work:
```
fwyzard@devfu-c2b04-44-01.cms:/data/user/fwyzard/CMSSW_12_5_0_pre5/src/HeterogeneousCore/AlpakaTest/test$ scram b runtests
>> Local Products Rules ..... started
>> Local Products Rules ..... done
Creating test log file logs/el8_amd64_gcc10/testing.log
Package HeterogeneousCore/AlpakaTest: Running test testHeterogeneousCoreAlpakaTestWriteRead
 
===== Test "testHeterogeneousCoreAlpakaTestWriteRead" ====
%MSG-i CUDAService:  (NoModuleName) 07-Sep-2022 22:14:53 CEST pre-events
CUDA runtime version 11.5, driver version 11.6, NVIDIA driver version 510.47.03
CUDA device 0: Tesla T4 (sm_75)
CUDA device 1: Tesla T4 (sm_75)
%MSG
%MSG-i AlpakaService:  (NoModuleName) 07-Sep-2022 22:14:53 CEST pre-events
AlpakaServiceCudaAsync succesfully initialised.
Found 2 devices:
  - Tesla T4
  - Tesla T4
%MSG
%MSG-i AlpakaService:  (NoModuleName) 07-Sep-2022 22:14:53 CEST pre-events
AlpakaServiceSerialSync succesfully initialised.
Found 1 device:
  - AMD EPYC 7763 64-Core Processor                
%MSG
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 07-Sep-2022 22:14:53.860 CEST
%MSG-i TestAlpakaAnalyzer:  TestAlpakaAnalyzer:testAnalyzerSerial  07-Sep-2022 22:14:53 CEST Run: 1 Event: 1
testTranscriberSerial.size() = 99
  data = 0x7f9ca572ba00,
  x    = 0x7f9ca572ba00,
  y    = 0x7f9ca572bd80,
  z    = 0x7f9ca572c100,
  id   = 0x7f9ca572c480,
  r    = 0x7f9ca572c680
  [y - x] = 0x380  [z - y] = 0x380  [id - z] = 0x380  [r - id] = 0x200
%MSG
...

07-Sep-2022 22:14:54 CEST  Initiating request to open file file:test.root
07-Sep-2022 22:14:54 CEST  Successfully opened file file:test.root
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 07-Sep-2022 22:14:54.814 CEST
%MSG-i TestAlpakaAnalyzer:  TestAlpakaAnalyzer:testAnalyzerSerial  07-Sep-2022 22:14:54 CEST Run: 1 Event: 1
testTranscriberSerial.size() = 99
  data = 0x7f208cf1b200,
  x    = 0x7f208cf1b200,
  y    = 0x7f208cf1b580,
  z    = 0x7f208cf1b900,
  id   = 0x7f208cf1bc80,
  r    = 0x7f208cf1be80
  [y - x] = 0x380  [z - y] = 0x380  [id - z] = 0x380  [r - id] = 0x200
%MSG
...
07-Sep-2022 22:14:54 CEST  Closed file file:test.root

---> test testHeterogeneousCoreAlpakaTestWriteRead succeeded
TestTime:2
^^^^ End Test testHeterogeneousCoreAlpakaTestWriteRead ^^^^
```

And it leaves behind the expected file:
```
fwyzard@devfu-c2b04-44-01.cms:/data/user/fwyzard/CMSSW_12_5_0_pre5/src/HeterogeneousCore/AlpakaTest/test$ edmDumpEventContent $CMSSW_BASE/testHeterogeneousCoreAlpakaTestWriteRead/test.root
Type                                  Module                    Label     Process    
-------------------------------------------------------------------------------------
PortableHostCollection<portabletest::TestSoALayout<128,false> >    "testProducer"            ""        "Writer"   
PortableHostCollection<portabletest::TestSoALayout<128,false> >    "testTranscriberSerial"   ""        "Writer"   
```

I can try using `&amp;&amp;` instead of `;`.

Reply:
ah sorry, I miss read the `command`, looks good.
Anyway, you need to replace `;` with `&amp;&amp;` otherwise scram is only going to get the exit code of last `cmsRun` command.

Reply:
The approach with `&amp;&amp;` does work, too:
```xml
<test
  name="testHeterogeneousCoreAlpakaTestWriteRead"
  command="
    mkdir -p testHeterogeneousCoreAlpakaTestWriteRead                            &amp;&amp;
    cd testHeterogeneousCoreAlpakaTestWriteRead                                  &amp;&amp;
    rm -f test.root                                                              &amp;&amp;
    echo                                                                         &amp;&amp;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py      &amp;&amp;
    echo                                                                         &amp;&amp;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py           &amp;&amp;
    echo                                                                         &amp;&amp;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py      &amp;&amp;
    echo                                                                         &amp;&amp;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py           &amp;&amp;
    echo                                                                         &amp;&amp;
    edmDumpEventContent test.root"
/>
```

However I found it a bit cumbersome, so I've tried using a subshell and setting `set -e` to make it abort on the first error:
```xml
<test
  name="testHeterogeneousCoreAlpakaTestWriteRead"
  command="(
    set -e;
    mkdir -p testHeterogeneousCoreAlpakaTestWriteRead;
    cd testHeterogeneousCoreAlpakaTestWriteRead;
    rm -f test.root;
    echo;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py;
    echo;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py;
    echo;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py;
    echo;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py;
    echo;
    edmDumpEventContent test.root;
  )"
/>
```

Unfortunately this does not work, because the whole `command` is being executed as part of a `command && ...` expression, which disabled the `-e` flag even in a subshell (see [this discussion](https://unix.stackexchange.com/questions/65532/why-does-set-e-not-work-inside-subshells-with-parenthesis-followed-by-an-or/254676#254676) and `man bash`).

So I'm considering a different approach, where the exit status of `cmsRun` is checked explicitly:
```xml
<test
  name="testHeterogeneousCoreAlpakaTestWriteRead"
  command="(
    mkdir -p testHeterogeneousCoreAlpakaTestWriteRead;
    cd testHeterogeneousCoreAlpakaTestWriteRead;
    rm -f test.root;
    echo;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py;
    echo;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/writer.py || exit $?;
    echo;
    echo cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py;
    echo;
    cmsRun ${LOCALTOP}/src/HeterogeneousCore/AlpakaTest/test/reader.py || exit $?;
    echo;
    edmDumpEventContent test.root;
  )"
/>
```
(the exit status of the final `edmDumpEventContent` is checked implicitly, because it is passed as the exit status of the whole subshell).

This seems to work as expected in all my tests, including when I forced the commands to to fail.

Reply:
By the way, yes, maybe I should just use a bash script to wrap the test...

Reply:
yes it is better and cleaner to wrap this in to a script ( which one can also directly run without `scram build runtests`). You can also do something like
```
<test name="testHeterogeneousCoreAlpakaTestWriteRead" command="testHeterogeneousCoreAlpakaTestWriteRead.sh">
  <flags USE_UNITTEST_DIR="1"/>
</test>
```
and scram will take care of creating `$CMSSW_BASE/unit_tests/testHeterogeneousCoreAlpakaTestWriteRead` directory and run the tests from that directory.